core/base/block_operator.cpp

```diff
@@ -0,0 +1,302 @@
+        for (size_type col = 0; col < blocks.front().size(); ++col) {
+            auto non_zero_col = find_non_zero_in_col(blocks, col);
+            if (blocks[row][col]) {
+                GKO_ASSERT_EQUAL_COLS(blocks[row][col], non_zero_col);
+                GKO_ASSERT_EQUAL_ROWS(blocks[row][col], non_zero_row);
+            }
+        }
+    }
+}
+
+
+template <typename Fn>
+std::vector<detail::value_span> compute_local_spans(
+    size_type num_blocks,
+    const std::vector<std::vector<std::shared_ptr<const LinOp>>>& blocks,
+    Fn&& get_size)
+{
+    validate_blocks(blocks);
+    std::vector<detail::value_span> local_spans;
+    size_type offset = 0;
+    for (size_type i = 0; i < num_blocks; ++i) {
+        auto local_size = get_size(i);
+        local_spans.emplace_back(offset, offset + local_size);
+        offset += local_size;
+    }
+    return local_spans;
+}
+
+
+dim<2> compute_global_size(
+    const std::vector<std::vector<std::shared_ptr<const LinOp>>>& blocks)
+{
+    validate_blocks(blocks);
+    if (blocks.empty()) {
+        return {};
+    }
+    size_type num_rows = 0;
+    for (size_type row = 0; row < blocks.size(); ++row) {
+        num_rows += find_non_zero_in_row(blocks, row)->get_size()[0];
+    }
+    size_type num_cols = 0;
+    for (size_type col = 0; col < blocks.front().size(); ++col) {
+        num_cols += find_non_zero_in_col(blocks, col)->get_size()[1];
+    }
+    return {num_rows, num_cols};
+}
+
+
+}  // namespace
+
+
+std::unique_ptr<BlockOperator> BlockOperator::create(
+    std::shared_ptr<const Executor> exec)
+{
+    return std::unique_ptr<BlockOperator>(new BlockOperator(std::move(exec)));
+}
+
+
+std::unique_ptr<BlockOperator> BlockOperator::create(
+    std::shared_ptr<const Executor> exec,
+    std::vector<std::vector<std::shared_ptr<const LinOp>>> blocks)
+{
+    return std::unique_ptr<BlockOperator>(
+        new BlockOperator(std::move(exec), std::move(blocks)));
+}
+
+
+BlockOperator::BlockOperator(std::shared_ptr<const Executor> exec)
+    : EnableLinOp<BlockOperator>(std::move(exec))
+{}
+
+
+BlockOperator::BlockOperator(
+    std::shared_ptr<const Executor> exec,
+    std::vector<std::vector<std::shared_ptr<const LinOp>>> blocks)
+    : EnableLinOp<BlockOperator>(exec, compute_global_size(blocks)),
+      block_size_(blocks.empty()
+                      ? dim<2>{}
+                      : dim<2>(blocks.size(), blocks.front().size())),
+      row_spans_(compute_local_spans(
+          block_size_[0], blocks,
+          [&](auto i) {
+              return find_non_zero_in_row(blocks, i)->get_size()[0];
+          })),
+      col_spans_(compute_local_spans(block_size_[1], blocks, [&](auto i) {
+          return find_non_zero_in_col(blocks, i)->get_size()[1];
+      }))
+{
+    for (auto& row : blocks) {
+        for (auto& block : row) {
+            if (block && block->get_executor() != exec) {
+                blocks_.push_back(gko::clone(exec, block));
```

Comment:
Is it necessary to clone these block operators ? The apply should handle that correctly anyway. I think there is no need to have all the blocks on the same executor.

Reply:
I'm following the same approach as `Composition` and `Combination`, which both do the same. I think it's better to be consistent here.